src/util/ecdsa.rs

```diff
@@ -417,6 +417,7 @@ impl<'de> ::serde::Deserialize<'de> for PublicKey {
 
 /// An ECDSA signature with the corresponding hash type.
 #[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```

Comment:
I am unsure this derive is particularly useful.

1. The fields are public so any consumer can easily take those two elements and stick them in a struct if they want to serialize them with serde.
2. `secp256k1::Signature` implements `Serialize` by means of serializing the DER-encoding.
3. This struct is just a bundle of how the signature ends up on the witness stack (where it is in serialized byte form).

If we wanted to provide a serde implementation for this, I'd vote for using the serialization format that ends up on the witness stack. However, given that `rust-bitcoin` doesn't use serde for consensus encoding / decoding, I don't think this `serde` implementation would be of much use.

Contrary, it might be confusing to take one of these and serialize it and suddenly end up with a struct when all you wanted is the actual signature (which is a field of this struct).

Given all this, I'd vote for not adding these to minimize confusion. What we should definitely have is a regression test for `secp256k1::Signature` but that should go into `rust-secp256k1`, not here I guess.

Reply:
I missed this. This was required for psbt serde support. I never had any confusion between serde encoding & consensus encoding, but I think this is a fair point.

> I'd vote for using the serialization format that ends up on the witness stack

I think there would many such instances of data structures in rust-bitcoin. I think rust-secp makes sure that serde represents how things end in the witness stack. I would prefer having the same witness serialization. but then we need to visit every auto derive in the library and do this for everything. And we lose the convenience of auto derives. 

At the very least right now, I think we should highlight that serde serialization is not the same as consensus serialization.

Reply:
#369 :)

Reply:
> And we lose the convenience of auto derives.

If all the fields are public, what is the value of providing these derives? If I need a data structure in my program that represents `EcdsaSig`, I can just add one myself and automatically get the freedom of choosing the field names I want etc.

I think we should seriously consider not providing serde implementations for these kind of container structs where there isn't a defined serialization format.

It triggers too many questions like "Is the defined serde format part of the public API?" Fixing a compile error because a field got renamed is one thing. Hitting a runtime deserialization error because a previously serialized struct changed a field name is much more annoying and invisible unless the end user adds explicit tests for it.

Additionally, for things like `Transaction`, I'd actually appreciate if we could expose a serde module that serializes to the consensus encoding: https://github.com/itchysats/itchysats/blob/58434af91d9692439a42f53f8b92a9d8b9193d72/daemon/src/model/cfd.rs#L1628-L1649

IMO, the need for such a module highlights that while choosing to derive `Serialize` might be convenient, it is also very opinionated. I think users would be better served by not implementing `Serialize` by default and instead provide multiple modules that users can use via `#[serde(with)]`:

- A module that serializes to consensus encoding like the one linked above
- A module that serializes to the internal structure, which can be created using serde's `remote` feature: https://serde.rs/remote-derive.html

Reply:
Yeah, changing (de)serialization is a breaking change and should be semver-breaking, mentioned in the changelog and people really should read changelogs before bumping versions.

Some things have obvious serializations in my mind: `Address`, `Amount`, `Txid`, `Blockhash` and probably some more.
`Transaction` has obvious binary serialization - the consensus one but non-obvious human serialization - should it be by fields like a struct or hex-encoded consensus?

Other things probably don't have obvious serialization especially if they are not transmitted often (at all) in practice. I don't remember seeing an application that transmits `EcdsaSig` alone. It's either a part of `Transaction` or a part of PSBT.